fe/fe-core/src/main/java/com/starrocks/connector/paimon/PaimonMetadata.java

```diff
@@ -292,4 +301,82 @@ public long getTableCreateTime(String dbName, String tblName) throws Exception {
         }
         return 0;
     }
+
+    public List<PartitionInfo> getChangedPartitionInfo(Table table, long latestSnapShotID) {
+        List<PartitionInfo> result = new ArrayList<>();
+        PaimonTable paimonTable = (PaimonTable) table;
+        Map<String, Long> partitionToSnapshotId = fetchChangedPartitionWithVersion(paimonTable, latestSnapShotID);
+        for (Map.Entry<String, Long> entry : partitionToSnapshotId.entrySet()) {
+            PaimonPartitionInfo partitionInfo = new PaimonPartitionInfo(entry.getKey(), entry.getValue());
+            result.add(partitionInfo);
+            if (entry.getValue() != null) {
+                latestSnapShotID = Math.max(latestSnapShotID, entry.getValue());
+            }
+        }
+        paimonTable.setLastedSnapshotId(latestSnapShotID);
+        LOG.debug("Get changed partitionInfo:{}", result);
+        return result;
+    }
+
+    private Map<String, Long> fetchChangedPartitionWithVersion(PaimonTable paimonTable, long nextSnapshotId) {
+        Map<String, Long> partitionToSnapshotId = new HashMap<>();
+        FileMonitorTable fileMonitorTable = new FileMonitorTable(paimonTable.getNativeTable());
+        Long latestId = fileMonitorTable.snapshotManager().latestSnapshotId();
+        long latestSnapshotId = latestId == null ? Long.MIN_VALUE : latestId;
+        if (nextSnapshotId >= latestSnapshotId) {
+            return partitionToSnapshotId;
+        }
+        ReadBuilder readBuilder = fileMonitorTable.newReadBuilder();
+        StreamTableScan scan = readBuilder.newStreamScan();
+        TableRead read = readBuilder.newRead();
+        if (nextSnapshotId != Long.MIN_VALUE) {
+            scan.restore(nextSnapshotId);
+        } else {
+            scan.restore(null);
+        }
+        try {
+            while (true) {
+                if (!scanMonitorTable(paimonTable, scan, read, partitionToSnapshotId)) {
```

Comment:
Is it possible to enter an infinite loop here?

I think it should return all partitions with latestSnapshotId if no splis exist in `fileMonitorTable` like catch exception below.